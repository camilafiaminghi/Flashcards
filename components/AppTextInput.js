import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { StyleSheet, View, Text, TextInput } from 'react-native'
import { validationRules } from '../utils/validation'
import { gray, black } from '../utils/colors'

class AppTextInput extends Component {

	static propTypes = {
		name: PropTypes.string.isRequired,
		placeholder: PropTypes.string.isRequired,
		maxLen: PropTypes.number.isRequired,
		onInputChange: PropTypes.func
	}

	state = {
		text: '',
		changed: false,
		valid: false
	}

	handleValidation = (value) => {
		const { name, onInputChange, maxLen } = this.props
		const valid = validationRules(name, value)

		if ( value.length <= maxLen ) {
			this.setState((state) => ({
				...state,
				text: value,
				changed: true,
				valid
			}))

			onInputChange(name, value, valid)
		}
	}

	render() {
		const { text, changed, valid } = this.state
		const { maxLen, placeholder } = this.props
		const textLeft = maxLen - text.length

		return (
			<View style={styles.container}>
				<TextInput
	        style={[styles.input, styles.inputText]}
	        placeholder={placeholder}
	        onChangeText={this.handleValidation}
	        value={text} />

	      { (textLeft !== 0 && textLeft < maxLen && changed) && <Text style={styles.text}>{ textLeft } characteres left</Text> }
	      { (textLeft === 0 && changed) && <Text style={styles.text}>max characters exceeded</Text>}
			</View>
		)
	}
}

export default AppTextInput

/* STYLES */
const styles = StyleSheet.create({
	container: {
		alignSelf: 'stretch',
		margin: 10
	},
	input: {
		flexDirection: 'row',
		alignSelf: 'stretch',
		height: 50,
		borderColor: gray,
		borderWidth: 1,
		borderRadius: 4
	},
	inputText: {
		padding: 8
	},
	text: {
		fontSize: 12,
		color: black
	}
})
