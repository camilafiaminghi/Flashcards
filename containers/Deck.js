import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { NavigationActions, HeaderBackButton } from 'react-navigation'
import { connect } from 'react-redux'
import { StyleSheet, Text, View, TouchableOpacity, Button } from 'react-native'
import { textColorInverse, textColor, pColor, pColorLight, pColorDark, sColor, sColorLight, sColorDark } from '../utils/colors'

class Deck extends Component {

	static navigationOptions = ({ navigation }) => {
		const { deckId } = navigation.state.params
		return {
			title: deckId,
			headerLeft: (
	      <HeaderBackButton
	        onPress={() => navigation.navigate('Decks', { parentNavigation: navigation })}
	      />
	    )
		}
	}

	static propTypes = {
		deck: PropTypes.object.isRequired,
		deckId: PropTypes.string.isRequired
	}

	toAddCard = () => {
		const { deckId } = this.props

		this.props.navigation.dispatch(NavigationActions.navigate({
			routeName: 'AddCard',
			params: { deckId }
		}))
	}

	toQuiz = () => {
		const { deck } = this.props

		this.props.navigation.dispatch(NavigationActions.navigate({
			routeName: 'Quiz',
			params: { deck }
		}))
	}

	render() {
		const { deck } = this.props

		return (
			<View style={styles.container}>
				<View style={styles.card}>
					<View style={styles.cardTitle}>
						<Text style={styles.text}>{deck.title}</Text>
						<Text style={styles.detail}>{deck.questions.length} { deck.questions.length === 1 && 'card' } { deck.questions.length !== 1 && 'cards' }</Text>
					</View>

					<TouchableOpacity
						style={[styles.input, styles.btn]}
						onPress={this.toAddCard}>
						<Text style={styles.btnText}>{'Add Card'.toUpperCase()}</Text>
					</TouchableOpacity>

					{ (deck.questions.length > 0) &&
						<TouchableOpacity
							style={[styles.input, styles.btn]}
							onPress={this.toQuiz}>
							<Text style={styles.btnText}>{'Start Quiz'.toUpperCase()}</Text>
						</TouchableOpacity>
					}
				</View>
			</View>
		)
	}
}

export const mapStateToProps = ({ decks }, { navigation }) => {
	const { deckId } = navigation.state.params

	return {
		deckId,
		deck: decks[deckId]
	}
}

export default connect(mapStateToProps)(Deck)

/* STYLES */
const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: sColorLight,
    paddingTop: 4,
    paddingBottom: 4
  },
  card: {
  	flex: 1,
  	flexGrow: 1,
  	alignItems: 'center',
  	justifyContent: 'center',
  	backgroundColor: 'white',
  	margin: 8,
		padding: 8,
  	borderRadius: 4,
  },
  cardTitle: {
  	marginBottom: 40,
  },
  text: {
  	textAlign: 'center',
  	fontSize: 26,
		marginRight: 24,
		marginLeft: 24,
		color: textColor
  },
  detail: {
  	alignSelf: 'center',
  	fontSize: 16,
  	color: sColorDark
  },
  input: {
		flexDirection: 'row',
		alignSelf: 'stretch',
		height: 50,
		borderBottomColor: pColorDark,
		borderBottomWidth: 1,
		borderRadius: 3,
	},
	btn: {
		alignItems: 'center',
		justifyContent: 'center',
		padding: 4,
		marginRight: 10,
		marginLeft: 10
	},
	btnText: {
		fontSize: 14,
		fontWeight: 'bold',
		color: pColorDark
	}
})
