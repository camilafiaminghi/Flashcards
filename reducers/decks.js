import {
		ADD_DECK,
		ADD_CARD } from '../actions'

const decks = (state = {}, action) => {
	switch (action.type) {
		case ADD_DECK:
			return {
				...state,
				[action.payload.title] : {
					title: action.payload.title,
					questions: []
				}
			}
		case ADD_CARD:
			return {
				...state,
				[action.payload.deckId] : {
					...state[action.payload.deckId],
					questions: [...state[action.payload.deckId].questions, action.payload.card]
				}
			}
		default:
			return state
	}
}

export default decks
