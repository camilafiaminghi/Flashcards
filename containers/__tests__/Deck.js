import React, { Component } from 'react'
import { Text, TouchableOpacity } from 'react-native'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import configureMockStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import { Deck, mapStateToProps } from '../Deck'
import AppTextInput from '../../components/AppTextInput'
import decks from '../../__mocks__/decks'

let wrapper
let instance
let component
const navigation = {
	navigate: jest.fn(),
	state: {
		params: {
			deckId: Object.keys(decks)[0]
		}
	},
	dispatch: jest.fn()
}
const props = {
	navigation,
	deckId: Object.keys(decks)[0],
	deck: decks[Object.keys(decks)[0]]
}

const configStore = configureMockStore([thunk])
const store = configStore({decks})

describe('<Deck />', () => {

	beforeEach(() => {
		wrapper = renderer.create(<Deck {...props} />)
		instance = shallow(<Provider store={store}><Deck {...props} /></Provider>)
		component = instance.find(Deck).shallow()
	})

	afterEach(() => store.clearActions())

	it('matches snapshot', () => {
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('should render', () => {
		expect(wrapper).toBeTruthy()
	})

	it('should has two Text components as children', () => {
		const title = component.find(Text).get(0)
		const questionsLen = component.find(Text).get(1)

		expect(title.props.children).toEqual(props.deck.title)
		expect(questionsLen.props.children).toEqual([props.deck.questions.length, ' ', false, ' ', 'cards'])
		expect(component.find(Text)).toHaveLength(4)
	})

	it('should has two TouchableOpacity components as children', () => {
		expect(component.find(TouchableOpacity).exists()).toEqual(true)
		expect(component.find(TouchableOpacity)).toHaveLength(2)
	})

	it('should handle toAddCard onPress', () => {
		const button = component.find(TouchableOpacity).get(0)
		button.props.onPress()
		expect(props.navigation.dispatch).toHaveBeenCalled()
	})

	it('should handle toQuiz onPress', () => {
		const button = component.find(TouchableOpacity).get(1)
		button.props.onPress()
		expect(props.navigation.dispatch).toHaveBeenCalled()
	})

	it('should mapStateToProps return props', () => {
		expect(mapStateToProps(store.getState(), { navigation })).toHaveProperty('deck')
	})
})
