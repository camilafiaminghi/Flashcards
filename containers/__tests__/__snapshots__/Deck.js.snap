// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Deck /> matches snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#efefef",
      "flex": 1,
      "paddingBottom": 4,
      "paddingTop": 4,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 4,
        "flex": 1,
        "flexGrow": 1,
        "justifyContent": "center",
        "margin": 8,
        "padding": 8,
      }
    }
  >
    <View
      style={
        Object {
          "marginBottom": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "fontSize": 26,
            "marginLeft": 24,
            "marginRight": 24,
            "textAlign": "center",
          }
        }
      >
        React
      </Text>
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#8d8d8d",
            "fontSize": 16,
          }
        }
      >
        2
         
         
        cards
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "stretch",
          "borderBottomColor": "#6f79a8",
          "borderBottomWidth": 1,
          "borderRadius": 3,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "marginLeft": 10,
          "marginRight": 10,
          "opacity": 1,
          "padding": 4,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#6f79a8",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        ADD CARD
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "stretch",
          "borderBottomColor": "#6f79a8",
          "borderBottomWidth": 1,
          "borderRadius": 3,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "marginLeft": 10,
          "marginRight": 10,
          "opacity": 1,
          "padding": 4,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#6f79a8",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        START QUIZ
      </Text>
    </View>
  </View>
</View>
`;
